torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,455 @@
+      "The requested batch size does not match with the initialized batch size.\n"
+      " The requested batch size is ", batch_size,
+      ", while the dataset is created with batch size equal to ", options_.batch_size_);
+
+    return batch_buffer_->get_batch();
+  }
+
+  /// This will clear any internal state and starts the internal prefetching
+  /// mechanism for the chunk dataset.
+  void reset() override {
+    // free workers from previous reset if there is any.
+    free_workers();
+    preload_threads_.clear();
+
+    chunk_reader_.reset();
+
+    size_t chunks_to_load = chunk_reader_.chunk_count();
+    chunk_sampler_.reset(chunks_to_load);
+
+    // Throw out any existing cached batch in the buffer and re-creates a new
+    // chunk buffer.
+    batch_buffer_ = torch::make_unique<
+        detail::BatchDataBuffer<UnwrappedBatchType, ExampleSamplerType>>(
+        chunks_to_load,
+        options_.batch_size_,
+        example_sampler_,
+        options_.ignore_empty_chunk_,
+        options_.cache_size_);
+
+    // create new workers for this new epoch.
+    quit_worker_ = false;
+
+    for (size_t i = 0; i < options_.preloader_count_; ++i) {
+      preload_threads_.emplace_back(
+          [this, i]() { this->preloader(i); });
+    }
+  }
+
+  /// size is not used for chunk dataset.
+  optional<size_t> size() const override {
+    return torch::nullopt;
+  }
+
+ private:
+  /// running on worker thread to preload chunk data.
+  void preloader(size_t id) {
+    while (!quit_worker_) {
+      try {
+        size_t chunk_id = 0;
+        if (auto chunk_sampler_result = chunk_sampler_.next(1)) {
+          chunk_id = chunk_sampler_result.value()[0];
+        } else {
+          break;
+        }
+        UnwrappedBatchType data = chunk_reader_.read_chunk(chunk_id);
+        AT_CHECK(options_.ignore_empty_chunk_ || !data.empty(),
+                 "Chunk with index " + std::to_string(chunk_id) + " is empty");
+
+        if (data.empty()) {
+          // skip adding the current chunk data and move to the next.
+          batch_buffer_->skip_chunk();
+        }
+        else {
+          batch_buffer_->add_chunk_data(std::move(data));
+        }
+      } catch (...) {
+        batch_buffer_->add_chunk_data(std::current_exception());
+      }
+    }
+  }
+
+  /// Block the current thread until the workers finish execution and exit.
+  void free_workers() {
+    if (!quit_worker_) {
+      quit_worker_ = true;
+      if(batch_buffer_){
+        batch_buffer_->stop();
+      }
+      for (auto& worker_thread : preload_threads_) {
+        worker_thread.join();
+      }
+    }
+  }
+
+ private:
+  // Templated class that defines what is a chunk and how to read chunk data.
+  // When a chunk is returned by chunk_reader_, ChunkDtatset split it into
+  // batches and caches them in batch_buffer_.
+  ChunkReader chunk_reader_;
+
+  // chunk sampler to shuffle different chunks
+  samplers::ThreadSafeSampler<ChunkSamplerType> chunk_sampler_;
```

Comment:
I'm a bit concerned that you're forcing the locking strategy here, while the provided sampler could have already been thread safe.

Reply:
To be honest it's better to add this later than remove it later, since the direction you're proposing is a breaking change (and the other one is not). That's my concern. Most samplers can easily be implemented using atomics and won't need any locks.

Reply:
We could have a `static_assert` that the `Sampler` someone templates on is a subclass of some dummy `ThreadSafeSampler` class, whose only purpose is to make people aware that the code they wrote should be made thread safe if they want to use it in this dataset.